awslabs/aws-lambda-powertools-typescript
# chore(logger): make Logger E2E works for all runtimes

## Description of your changes

Currently, E2E only runs on a single (default) runtime from CDK Lambda module. With this change,`test:e2e` target runs tests in all runtime to detect any bugs from compatibility issue. It will create 1 stack of test Lambda functions per run time and execute the tests in parallel.

NodeJS10 has been deprecated. So we will be running on v12 and v14 for now.

### Why this approach?
There are several options to achieve this. Here are the main criteria in selecting an approach.
1. **Speed** Able to deploy infrastructure (in `beforeAll()` block) concurrently.
1. **Cognitive load for test writers** Test writers can focus more on testing and don't get distracted by having to think of multiple runtimes. 
1. **Easy to read the log** Logs from concurrent test runs (from multiple runtimes) will overlap with each other. 

Here are the list of options I evaluated:
1. **Use `it.each()` and `describe.each()` to iterate an array of runtimes.** ([link](https://dev.to/bgord/simplify-repetitive-jest-test-cases-with-test-each-310m)) This option doesn't work as Jest won't parallel the `beforeAll()` execution in the same file.  It doubles the time to run the tests. In addition, test writers have to explicitly specify an array of run-time, more boilerplate code.
2. **Have a custom script `run-e2e.ts` that starts subprocess to run concurrently** . This works with Node's `child_process`'s  `exec`. However, the log coloring is lost. In additional, we need to maintain an additional script.
3. **Use `forEach` to iterate `describe` blocks to run each runtime test independently**. This also won't run `beforeAll()` in parallel. Jest that decides to run things in parallel or sequential. We have no control over it.   describe and it basically put things in an internal queue and jest has [some heuristic algorithm to decide](https://github.com/facebook/jest/issues/5818#issuecomment-375950546).  I find that it stops running in parallel when when several describe are called in the same file.
4. **Use [concurrently](https://www.npmjs.com/package/concurrently) and pass runtime as an environment variable.** I find this option cleanest and satisfy the 3 criteria above. The logs overlap with each other but we there is a prefix stating which runtime the log comes from.  

### How to verify this change
Check out this branch and run
```sh
cd packages/logger
AWS_PROFILE=<YOUR_PROFILE> AWS_REGION=eu-central-1 npm run test:e2e
```

Result from local run
```
[test:e2e:nodejs12x]
[test:e2e:nodejs12x] Test Suites: 3 passed, 3 of 8 total
[test:e2e:nodejs12x] Tests:       13 passed, 13 total
[test:e2e:nodejs12x] Snapshots:   0 total
[test:e2e:nodejs12x] Time:        115.722 s, estimated 136 s
[test:e2e:nodejs12x] Ran all test suites.
[test:e2e:nodejs12x] npm run test:e2e:nodejs12x exited with code 0
[test:e2e:nodejs14x]
[test:e2e:nodejs14x] Test Suites: 3 passed, 3 of 8 total
[test:e2e:nodejs14x] Tests:       13 passed, 13 total
[test:e2e:nodejs14x] Snapshots:   0 total
[test:e2e:nodejs14x] Time:        115.771 s, estimated 136 s
[test:e2e:nodejs14x] Ran all test suites.
[test:e2e:nodejs14x] npm run test:e2e:nodejs14x exited with code 0
```

### Related issues, RFCs
#353

### PR status

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

## Checklist


### Breaking change checklist


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
